toolchain/check/testdata/interop/cpp/impls/destroy.carbon

```diff
@@ -0,0 +1,256 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/convert.carbon
+// EXTRA-ARGS: --clang-arg=--std=c++20
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/interop/cpp/impls/destroy.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/interop/cpp/impls/destroy.carbon
+
+// --- types.h
+
+class PublicDestructor {
+ public:
+  ~PublicDestructor();
+};
+
+class TrivialDestructor {
+ public:
+  ~TrivialDestructor() = default;
+};
+
+class DeletedDestructor {
+ public:
+  ~DeletedDestructor() = delete;
+};
+
+template<int N = 0>
+class AmbiguousDestructorT {
+ public:
+  ~AmbiguousDestructorT() requires (N != 1);
+  ~AmbiguousDestructorT() requires (N != 2);
+};
+
+using AmbiguousDestructor = AmbiguousDestructorT<>;
+
+class PrivateDestructor {
+ private:
+  ~PrivateDestructor();
+};
+
+class ProtectedDestructor {
+ protected:
+  ~ProtectedDestructor();
+};
+
+// --- destroy_destroyable.carbon
+
+library "[[@TEST_NAME]]";
+
+import Cpp library "types.h";
+
+fn PublicDestroy() {
+  //@dump-sem-ir-begin
+  var a: Cpp.PublicDestructor = {};
+  //@dump-sem-ir-end
+}
+
+fn TrivialDestroy() {
+  //@dump-sem-ir-begin
+  var a: Cpp.TrivialDestructor = {};
+  //@dump-sem-ir-end
+}
+
```

Comment:
Do we want a test where we inherit from ProtectedDestructor in Carbon and destroy it?

Reply:
Done. That works, but deriving from `PrivateDestructor` also works, so I've added a `todo_fail` test for that.